### DISCUSSION TITLE
Welcome to RosieLIS Discussions!

### DISCUSSION THREAD

**Discussion by roy-harmon** [roy-harmon/RosieLIS#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by muneebz-jeevaysoft** [comment]
Does the Universal LIS also support Siemens Dimension ?

**Reply by roy-harmon** [reply]
No, the Dimension line doesn't use the ASTM/CLSI communication standard, so it is not supported by UniversaLIS.

**Reply by muneebz-jeevaysoft** [reply]
Thank you for the prompt response , do we have any LIS built by any chance for Siemens Dimension or similar protocol machines through TCP/IP , I am new to this , so looking for something I can go over and build on

**Reply by roy-harmon** [reply]
You could start with RosieLIS, change the serial port to a TCP socket, and start from there. With the knowledge I gained from working on UniversaLIS, I can probably add it to this program when I get a chance. It may not be in the next few days, but within the next month or so, I'll try to make it happen.

**Reply by muneebz-jeevaysoft** [reply]
Thank you so much, Is it possible for you to add TCP socket to Rosie LIS , I have a python background, so struggling a little with VB.Net

**Reply by roy-harmon** [reply]
I've taken a look at the interface specification document in the Siemens Document Library (https://doclib.siemens-healthineers.com/rest/v1/view?document-id=600706) and I'm not very confident that TCP/IP support is feasible. The specification states: 
> If an instrument is connected to a Siemens automation or middleware product, the instrument is typically set up to communicate using TCP/IP protocol. If the instrument is not connected directly to a Siemens automation or middleware product, it must be connected using serial protocol.

Furthermore, the operator's manuals for both the Dimension XPand Plus and the Dimension EXL indicate that the 9-pin (DB9) serial port connection is required for host interfacing:
> For host interfacing, a 9-pin female connector is required for hookup to the male connector on the Dimension® EXL™ system host communications port.

Considering that there is no other mention of TCP/IP protocol anywhere else in the interface specification and no information about the instrument's ethernet port in the operator's manual, I suspect this option is exclusive to Siemens middleware installed by Siemens field service technicians. In that case, I suspect the data transmission does not follow the same protocol. It's entirely possible that data transferred by Siemens automation/middleware bypasses the Dimension software's LIS module altogether. 

Is there a specific reason why you're looking to use TCP instead of a serial port? If the instrument is too far away to run a serial cable or your host computer has no RS232 serial port, you may have more luck with one of the various "serial to ethernet" adapters available online for around $50. I haven't tried any of them, but if they work as advertised, I suspect that would be a cheaper alternative than whatever Siemens charges for their middleware solution.

**Comment by muneebz-jeevaysoft** [comment]
Hi Roy, Yes what you are saying is absolutely correct. But the LIS server we would be hosting will be not i side the LAB , but outside sitting in a small AWS machine, As the Lab does not have servers in the lab to support it there. What we have managed is using a lantronix box , what it does is that it is connected to COM port , and it forwards the data to the TCP IP and Port we provide. We were able to receive the data using a test tcp server from the siemens machine. This is the reason we want to go with TCP IP

**Reply by roy-harmon** [reply]
Have you already tried using the RosieLIS program by setting up a virtual COM port on the machine that hosts your LIS? I understand that you want to communicate directly by TCP, but until I modify the program to work with third-party hardware, that might be a good alternative. 
Lantronix appears to have a program called Com Port Redirector to do that, available here: https://ltrxdev.atlassian.net/wiki/spaces/LTRXTS/pages/106070498/Latest+Version+of+Lantronix+CPR+Com+Port+Redirector

**Reply by muneebz-jeevaysoft** [reply]
This sound like a good workaround, let me see if that will be possible. Meanwhile thanks for the help

**Comment by muneebz-jeevaysoft** [comment]
Hi @roy-harmon   , I was able to try VCOM and receive the data , thanks for the valuable suggestion

Until now I was just installing the service using the LIS_Setup.msi you provided , but now as I see few errors and want to change code a little and add few more logs and maybe modify the queries to adjust to the setup we have , but when I make changes and build the project , it generates a RoyHarmon.RosieLIS.dll.exe file in Debug folder and when I install that file as a service and run it, It shows running but does not log anything ,I also tried the debugger and it does not reach the OnStart function of the RosieLis class , but still shows running as a service 

This is the first project for me in VB.net or Visual Studio, kindly bear with me :'(

**Reply by roy-harmon** [reply]
It's been a while since I worked with the installer. I'll have to take a look at this in more detail, probably this weekend.

In the meantime, have you created your own fork of this repository? That might make troubleshooting easier since I would be able to look at your code.

**Reply by muneebz-jeevaysoft** [reply]
To be honest I did not do a lot of changes , but when I build the project and try with the new exe it does not seem to work and does not log anything in the file under allusersprofile. When I tried to debug it and added a breakpoint under the main the class , I see the partial class name is RosieLIS and the below line

ServicesToRun = New System.ServiceProcess.ServiceBase() {New RosieLIS}

It executes , but not sure which onStart function it is calling as under RosieLIS.vb , the name of class is RosieLISService, The onStart of RosieLISService is never invoked

**Reply by roy-harmon** [reply]
Did you build the LIS_Setup project and launch the LIS_Setup.msi installer from your RosieLIS\LIS_Setup\Release or RosieLIS\LIS_Setup\Debug folder? That seems to be working for me.

**Reply by muneebz-jeevaysoft** [reply]
@roy-harmon  , which version of visual studio are you using to build this project ?

**Reply by roy-harmon** [reply]
Visual Studio 2022 Community Edition.

**Reply by muneebz-jeevaysoft** [reply]
Yes the the msi file is created in DEBUG folder and I am to install it as well , but when I install this newly built MSI , I do not see anything logged in the SerialLog_ file when the service is started. But when I use the MSI that was created 4 years ago under the LIS_SETUP , and install that MSI , I can see the logs. The Event log does not show any error as well. Were you able to see new logs generated after you ran with the newly built MSI ? @roy-harmon

**Reply by roy-harmon** [reply]
I don't currently have a system set up to test it with, so without any data coming in, no log file is created. I do see the entry in the Event Viewer (Windows Logs > Application) for RosieLIS saying "Service started successfully." I'll have to look into what's going on here.

**Reply by muneebz-jeevaysoft** [reply]
@roy-harmon  , for example I am using this in my local system as well and dont have any com port currently. But with old LIS setup file, I still get a log saying COM port not found, but when I build and run the LIS setup file from Debug folder, I see its running , but It does not log any error saying COM port not found like the old setup file. Even if I user AppendToLog function during the starting of the OnStart function as well, I do mot see that log created. Sorry, for buzzing you continuosly, but trying to understand what issue might be when the Lis setup is being built.

**Reply by roy-harmon** [reply]
Yeah, I see what you mean. I'll work on it and see if I can figure out the problem. I have a virtual machine that I should be able to hook up to test it.

The hardest part of writing this program has always been the installer. Installing the Windows service is very tricky, and it's only gotten harder as time goes on. The newer frameworks have made it even more difficult, but this is still on the old .NET Framework, I should be able to get this one working.

**Reply by muneebz-jeevaysoft** [reply]
Thanks a lot @roy-harmon   , Appreciate it , your support and timely responses have been tremendous help to me.

**Reply by roy-harmon** [reply]
I've found at least one issue so far: the System.IO.Ports dependency has dropped support for .NET Framework 4.5.2 and requires at least 4.6.1. I don't know why this wasn't caught by Visual Studio during the build process, but I think it's only one of the problems I'll need to solve. 

I'll keep digging and let you know what I find.

**Reply by muneebz-jeevaysoft** [reply]
That's great news, let me know if the latest build works for you, if it does, I will try cloning the project and check it at my end as well. @roy-harmon

**Reply by muneebz-jeevaysoft** [reply]
Hi @roy-harmon  , were you able to see if there were any more issues ?

**Reply by roy-harmon** [reply]
Yes, but after working with it for several hours, I still haven't been able to get it to work properly. At this point I think something in the solution configuration must have been corrupted in the migration from VS2015. I'll probably just need to create a new project in VS2022 and import the necessary modules. This will probably help me clean out some of the leftover metadata clutter, too.

**Reply by muneebz-jeevaysoft** [reply]
Thanks @roy-harmon  , Appreciate the help , will be waiting for the changes

**Reply by roy-harmon** [reply]
I believe I've fixed it now. Try it out and let me know what you think. I also modified it so you should be able to debug it easily from the IDE now.
https://github.com/roy-harmon/RosieLIS/releases/tag/v1.2.0

**Reply by muneebz-jeevaysoft** [reply]
Hi @roy-harmon , thank you so much , the debug is working fine now and I can see the log is generated , but the service fails to be in running mode and stops , Meaning it starts after few seconds it stops. I checked and recieved the logs that the Com port is connected and the DB is connected as well. But when I debugged actually I see the code only runs the OnStart function and it stops, It does not wait for the com data to be received, can you advice

**Reply by roy-harmon** [reply]
All right, I'll see what I can do with that. I've got a lot on my plate at the moment, so it might take me a week or two.

In the meantime, feel free to tinker with it. My ability to test it is limited at the moment, so you might be able to fix it better than I can.

**Reply by roy-harmon** [reply]
I took out the line that made the service stop when debugging. See what it does now.

**Reply by muneebz-jeevaysoft** [reply]
Hi @roy-harmon  sadly the behavior is still the same as the OnStart function runs and open comport and sql connnection code is run successfully but after that the code exits , what I am trying to see is the service should not be stopping , The LIS_Setup.msi file (4 years old) , continuously runs , even if it fails to open the com port , it keeps on retrying , but here it just runs once, I am trying to figure out why that is happening
